### ISSUE TITLE
Set due date can confuse FSRS's calculations

### ISSUE THREAD

**Issue by dae** [ankitects/anki#3634]
> After updating to Anki 24.11, and re-optimising my parameters, some of my cards have had their intervals increased by an extremely large amount of time, for no apparent reason. For example, in the image below, you can see a card that before I rescheduled had an interval of 27 days. However, after updating to Anki 24.11 (and so FSRS 5), and rescheduling my cards, the interview changed to 6.67 years. I most definitely do not know this card, and a select few other cards that have had similar changes to their intervals.

Originally reported on https://forums.ankiweb.net/t/fsrs-5-increased-interval-by-5-years/53003/7

**Comment by dae** [comment]
@L-M-Sherlock Would the previously-discussed storing of last revlog timestamp in the card data (partially) resolve this?

**Comment by L-M-Sherlock** [comment]
Nope. It's another issue. The problem is we convert the card's current interval to stability again and again.

**Comment by user1823** [comment]
> Then Anki cannot find the first_of_last_learn_entries
> So Anki infers the stability from card’s interval

Shouldn't Anki infer the stability from the interval of the first usable revlog?

That is, from the interval of the entry dated 20th May 2024 in the below screenshot.
<img width=600 src="https://github.com/user-attachments/assets/a3a4f034-812b-43cf-8cc5-08f1c3ad56b5">

**Comment by L-M-Sherlock** [comment]
2024-03-15 is an revlog entry generated by `Reset`. So the previous entries are ignored.

**Comment by user1823** [comment]
I said 20th May

**Comment by L-M-Sherlock** [comment]
The type of 20th May is review, instead of learn.

**Comment by user1823** [comment]
Yes, but Anki infers memory states from `Review` type of revlogs (using memory_state_from_sm2) in case of incomplete revlogs. Right?

**Comment by L-M-Sherlock** [comment]
At least the current code doesn't.

**Comment by user1823** [comment]
In some cases, it does. Maybe this case doesn't match the condition placed in the code?

**Comment by L-M-Sherlock** [comment]
If Anki doesn't find a `Learn` entry after the `Forget` entry, all of entires are ignored:

https://github.com/ankitects/anki/blob/d90050600316cd58d25bc7f61fc762c3f704c101/rslib/src/scheduler/fsrs/params.rs#L276-L286

https://github.com/ankitects/anki/blob/d90050600316cd58d25bc7f61fc762c3f704c101/rslib/src/scheduler/fsrs/memory_state.rs#L321-L322

https://github.com/ankitects/anki/blob/d90050600316cd58d25bc7f61fc762c3f704c101/rslib/src/scheduler/fsrs/memory_state.rs#L185-L191

**Comment by user1823** [comment]
You added this code in https://github.com/ankitects/anki/commit/2b890b0eee7a9fb40a137ec2ce7aacd9044d0258 to fix the problem that Anki was using new (=reset) cards for optimization.

Instead of checking whether there is a learn entry or not, doesn't it make sense to check if there is any entry after the Reset entry?

Here is my proposed pseudo-code:

```rs
 ) { 
     // If we find a `Learn` entry after the `Forget` entry, we should 
     // treat the revlogs as complete and ignore the entries before the `Forget` entry 
     if first_of_last_learn_entries.is_some() { 
         revlogs_complete = true;    (WHY was this false in the earlier code? Shouldn't it be true?)
         break; 
     // If we find `Review` entries but not a `Learn` entry after the `Forget` entry , it's 
     // treat the revlogs as incomplete and ignore the entries before the `Forget` entry 
     } elseif non_manual_entries.is_some() { 
         revlogs_complete = false;
         break; 
     // If we don't find any non-manual entry after the `Forget` entry, it's 
     // a new card and we should ignore all entries 
     } else { 
         return None; 
     } 
```

**Comment by L-M-Sherlock** [comment]
Looks good! I will draft up a PR to solve it by your method.

**Comment by L-M-Sherlock** [comment]
It looks better:

![Image](https://github.com/user-attachments/assets/1fdf8101-2efd-4061-af30-0be732c09a98)
